src/Haskell/Ide/Engine/Dispatcher.hs

```diff
@@ -47,17 +48,31 @@ dispatcherP :: forall m. TChan (PluginRequest m)
             -> DispatcherEnv
             -> ErrorHandler
             -> CallbackHandler m
+            -> J.ClientCapabilities
             -> IO ()
-dispatcherP inChan plugins ghcModOptions env errorHandler callbackHandler =
-  void $ runIdeGhcM ghcModOptions (IdeState emptyModuleCache Map.empty plugins Map.empty Nothing) $ do
-    stateVar <- lift . lift $ ask
-    gchan <- liftIO $ do
-      ghcChan <- newTChanIO
-      ideChan <- newTChanIO
-      _ <- forkIO $ mainDispatcher inChan ghcChan ideChan
-      _ <- forkIO $ runReaderT (ideDispatcher env errorHandler callbackHandler ideChan) stateVar
-      return ghcChan
-    ghcDispatcher env errorHandler callbackHandler gchan
+dispatcherP inChan plugins ghcModOptions env errorHandler callbackHandler caps =
+  void $ runIdeGhcM ghcModOptions caps (IdeState emptyModuleCache Map.empty plugins Map.empty Nothing) $ do
+      stateVar <- lift $ lift $ lift ask
+      gchan <- liftIO $ do
+        ghcChan <- newTChanIO
+        ideChan <- newTChanIO
+
+        _ <- forkIO $ mainDispatcher inChan ghcChan ideChan
+        _ <- forkIO $ flip runReaderT stateVar $
+                        flip runReaderT caps $ 
+                          ideDispatcher env errorHandler callbackHandler ideChan
+
+        -- TODO: Causes STM deadlock?
```

Comment:
What is going on here with this comment?

Reply:
I tried to get the ideDispatcher thread to throw its exceptions onto the main thread instead of dying silently, but it caused other exceptions due to me not doing the async properly. We should probably revisit this some time. I can remove the comment and create a todo for this?